### ISSUE TITLE
HA 2024.2.0 and 2024.2.1 - unsupported operand type(s) for |: 'list' and 'set'

### ISSUE THREAD

**Issue by megosugit** [petretiandrea/home-assistant-tapo-p100#694]
## Version of the integration

2.13.0 (latest)

## Configuration

L920-5
Firmware: 1.1.3 Build 231229 Rel.164316

```
unsupported operand type(s) for |: 'list' and 'set'
```

## Describe the bug

Since I upgraded my HA from 2024.1.5 to 2024.2.0 (same with 2024.2.1), I cannot turn off my Smart Light Strip.
Turning on works, but the device stays Off in HA.
Turning it off gives the following error:
unsupported operand type(s) for |: 'list' and 'set'

I read in the HA changelogs that Tapo is now directly integrated without HACS. But I can't find any documentation about the transition.
What should I do to fix it ?

FYI: P115 on/off is still working fine.

## Debug log

```
TypeError: unsupported operand type(s) for |: 'list' and 'set'
2024-02-08 15:26:59.340 INFO (MainThread) [custom_components.tapo.light] Setting brightness: None, tapo None
2024-02-08 15:26:59.340 INFO (MainThread) [custom_components.tapo.light] Setting color: None, tapo None
2024-02-08 15:26:59.340 INFO (MainThread) [custom_components.tapo.light] Setting color_temp: None, tapo None
2024-02-08 15:26:59.340 INFO (MainThread) [custom_components.tapo.light] Setting effect: None
2024-02-08 15:26:59.466 DEBUG (MainThread) [custom_components.tapo.coordinators] Finished fetching tapo data in 0.089 seconds (success: True)
2024-02-08 15:26:59.466 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [140021222090688] unsupported operand type(s) for |: 'list' and 'set'
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 240, in handle_call_service
    response = await hass.services.async_call(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2279, in async_call
    response_data = await coro
                    ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/core.py", line 2316, in _execute_service
    return await target(service_call)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 892, in entity_service_call
    single_response = await _handle_entity_call(
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 962, in _handle_entity_call
    result = await task
             ^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/light/__init__.py", line 628, in async_handle_light_on_service
    await light.async_turn_on(**filter_turn_on_params(light, params))
  File "/config/custom_components/tapo/light.py", line 166, in async_turn_on
    await self.coordinator.async_request_refresh()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 263, in async_request_refresh
    await self._debounced_refresh.async_call()
  File "/usr/src/homeassistant/homeassistant/helpers/debounce.py", line 93, in async_call
    await task
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 289, in async_refresh
    await self._async_refresh(log_failures=True)
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 412, in _async_refresh
    self.async_update_listeners()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 183, in async_update_listeners
    update_callback()
  File "/config/custom_components/tapo/entity.py", line 43, in _handle_coordinator_update
    self.async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 989, in async_write_ha_state
    self._async_write_ha_state()
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1110, in _async_write_ha_state
    state, attr, capabilities, shadowed_attr = self.__async_calculate_state()
                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 1049, in __async_calculate_state
    attr.update(self.state_attributes or {})
                ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/light/__init__.py", line 1129, in state_attributes
    self.__validate_color_mode(color_mode, legacy_supported_color_modes, effect)
  File "/usr/src/homeassistant/homeassistant/components/light/__init__.py", line 1093, in __validate_color_mode
    effect_color_modes = supported_color_modes | {ColorMode.ONOFF}
                         ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~

```

**Comment by jcbrownratby** [comment]
similar  for me. started at the time the TP-Link Smart Home became availabe in HA (2024.2) i think)

if i select a colour it wokrs fine, but if i select an effect i get the error . i get the errir at the point i select the effect and when i then press on or off. As the "TP-Link Smart Home" integration does not support effects on the L920-5, i'm sort of stuck

**Comment by jcbrownratby** [comment]
I appear to have found a way round this. I'm not saying it is a fix, or even the right thing to do, but it works for me, and i have seen no ill-effects from it so far.

- use whatever mechanism you prefer (WinSCP, terminal to HA, Studio Code Server, etc) and navigate to "/<HA>/config/custom_components/tapo/"
- Make a backup copy of the "const.py" file, just in case you mess up the editing in the next steps
- open the "const.py" in your prefered text editor.
- locate the SUPPORTED_LIGHT_EFFECTS fo the l920 (and l930 and l900) - in my system that was lines 53 - 57

```
      SUPPORTED_LIGHT_EFFECTS = [
          "l930": LightEffectPreset,
         "l920": LightEffectPreset,
         "l900": LightEffectPreset,
      ]
```

- Change the square brackets on lines 53 and 57 to curly brackets:
```
      SUPPORTED_LIGHT_EFFECTS = {
          "l930": LightEffectPreset,
         "l920": LightEffectPreset,
         "l900": LightEffectPreset,
      }
```

- save and close the file

- I then rebooted my HA. no idea if this is needed or not, but i thought i would do it to be sure.

I am not a python programmer ( I am a programmer in other languages, but not Python), but found some information by Googling this error, and looking at the code.

Happy to be shot down in flames if this is the wrong thing to do.

**Comment by megosugit** [comment]
Hi,
I looked in my setup, I already have the curly brackets, but it doesn't work well.
My const.py shows VERSION = "2.13.0", what about yours ?

Yesterday, I thought it was fixed after upgrading to HA 2024.2.2, I could switch on/off from HA, but later on, when I tried to switch it off, it was showing as already off in HA (it wasn't in real life nor in Tapo app), from there I got the same error as usual. So nothing new for me :(

Do you think there is a way to manage white temperature with this strip ? I only have color slider in HA, but no white cold/warm slider.
Same in Tapo app.

**Comment by jcbrownratby** [comment]
just realised, I had changed something else in const.py, see updated notes above - sorry about that.

Yes, I am on the same version in const.py (2.13.0), and HA 2024.2.2, and this change does work for me. 

like you, before my hack, if i tried to turn the strip on as a  constant colour, everything was ok.. if i selected an effect I got the error.. the light did (usually) change to the effect I had selected, but it did not change in HA (or the Tapo App) as being shown switched on.

after my hack HA turns on the light, with the selected effect, an no error

Same for me, no colour temp in HA. I have tried adding the ColorMode.COLOR_TEMP to the definition for the l920 AND woo woo, i get he temperature settiing in HA. BUT...if i try to use this i ger an error :(

**Comment by megosugit** [comment]
it seems to work fine with:
"l920": {ColorMode.ONOFF, ColorMode.BRIGHTNESS, ColorMode.HS}, 

I'll continue to use it with your changes, and see if it continues working fine.
Great find !

I wonder why nobody else has reported this issue and why it is left unanswered ¿

**Comment by jcbrownratby** [comment]
Glad you have got it working :)

I relooked into the colour temp, and i had misread your comment. I dont think the l920 supports setting the colour temp, which is why there is no facility either in teh TAPO App or in HA.. so when I add it to HA then yes it displays the option (as I am now telloing HA that colour temp is supported for the l920. But of course, when i try to use it in HA, then it fails - HA is asking Tapo l920 to set the color temp. TAPO is saying back "You want me to do what?... Im an l920, and i cannot do that" (or similar errir :) ), so i then get an errr in HA.

So,i'm guessing that  unles TPLink provide some sort of firmware update for the l920 which supports colour-temp, (of that is even a possibility with this strip), then HA (nor anything) will not be able to set the colour temp on an l920 strip

**Comment by megosugit** [comment]
What I find curious, is the option is possible in Tapo app.
On the main screen, you only have the brightness and color control, but if you edit one of the 7 presets in Tapo app, then you can chose "White light" and you get a temperature slider to change the Kelvin.

**Comment by petretiandrea** [comment]
Fixed on 2.14.0
